mesonbuild/dependencies/base.py

```diff
@@ -505,51 +505,61 @@ def __init__(self, name, environment, kwargs, language=None):
-            else:
-                mlog.debug('Pkg-config binary for %s is cached.' % for_machine)
-                choice = PkgConfigDependency.class_pkgbin[for_machine]
-                assert choice is not None
-                return choice
             # Lookup in cross or machine file.
-            bt = environment.binaries[for_machine]
-            potential_pkgpath = bt.lookup_entry('pkgconfig')
-            if potential_pkgpath is None:
-                mlog.debug('Pkg-config binary missing from cross or native file, or PKG_CONFIG undefined.')
-            else:
-                mlog.debug('Pkg-config binary for %s specified from config file as %s.', for_machine, potential_pkgpath)
-                potential_pkgbin = ExternalProgram.from_entry('pkgconfig', potential_pkgpath)
-                if not potential_pkgbin.found():
-                    mlog.debug(
-                        'Pkg-config %s for machine %s specified at %s but not found.',
-                        potential_pkgbin.name, for_machine, potential_pkgbin.command)
-                else:
-                    return potential_pkgbin
+            potential_pkgpath = environment.binaries[for_machine] \
+                .lookup_entry('pkgconfig')
+            if potential_pkgpath is not None:
+                mlog.debug('Pkg-config binary for %s specified from cross file, native file, or env var as %s.', for_machine, potential_pkgpath)
+                yield ExternalProgram.from_entry('pkgconfig', potential_pkgpath)
+                # We never fallback if the user-specified option is no good, so
+                # stop returning options.
+                return
+            mlog.debug('Pkg-config binary missing from cross or native file, or env var undefined.')
             # Fallback on hard-coded defaults.
+            # TODO prefix this for the cross case instead of ignoring thing.
             if environment.machines.matches_build_machine(for_machine):
                 for potential_pkgpath in environment.default_pkgconfig:
-                    potential_pkgbin = self.check_pkgconfig(potential_pkgpath)
-                    if potential_pkgbin is None:
-                        mlog.debug(
-                            'default Pkg-config fallback %s for machine %s specified at %s but not found.',
-                            potential_pkgbin.name, for_machine, potential_pkgbin.command)
-                    else:
-                        return potential_pkgbin
-
-        self.pkgbin = search()
-        if self.pkgbin is None:
-            msg = 'Pkg-config binary for machine %s not found.' % for_machine
+                    mlog.debug('Trying a default pkg-config fallback at', potential_pkgpath)
+                    yield ExternalProgram(potential_pkgpath, silent=True)
+
+        # Only search for pkg-config for each machine the first time and store
+        # the result in the class definition
+        if PkgConfigDependency.class_pkgbin[for_machine] is False:
```

Comment:
`if not PkgConfigDependency...`

Reply:
I prefer to be explicit because `None` and `False` are separate here.

Reply:
Okay, I guess that makes sense.